[READY FOR REVIEW] docker image, for standalone or quick-launch

ahgittin opened the pull request:
see the README

super-easy way to launch brooklyn or even package a special brooklyn for a specific blueprint

everything seems to be working except the HL blueprint needs to refer to the single cluster example

note this is already pushed to docker hub so you can experiment on the runs without checking anything out :)

@grkvlt @mikezaccardo @shartzel welcome any comments

m4rkmckenna commented on docker/Dockerfile:
```diff
@@ -0,0 +1,68 @@
+RUN if [ -n "${debug}" ] ; then apt-get install -y zip tar less curl ; fi
+RUN if [ -n "${debug}" ] ; then apt-get install -y sudo ; fi
+
+RUN adduser brooklyn
```
[Good practice](https://docs.docker.com/engine/userguide/eng-image/dockerfile_best-practices/) is to not add multiple fs layers 

Why not combine all the run commands into one
```
RUN if [ -n "${debug}" ] ; then apt-get install -y vim ; fi \
        if [ -n "${debug}" ] ; then apt-get install -y zip tar less curl ; fi \
       if [ -n "${debug}" ] ; then apt-get install -y sudo ; fi \
       adduser brooklyn
```

ahgittin commented:
great comments -- will update to use alpine and cut down fs layers

ahgittin commented:
@andreaturli @m4rkmckenna @mikezaccardo fixed HL issue, and addressed comments.  6a47cf2 shows the difference in filesize w alpine (huge!)

updating issue title; ready for review

re https://github.com/fabric8io/docker-maven-plugin the makefile is there as documentation more than anything.  it would be nice to have this part of the build but it would require docker be installed in the target system so i've held off for now.

cool things to try...

deploy **hyperledger blockchain**:

    docker run -d -p 8081:8081
      brooklyncentral:brooklyn-hyperledger \
      --username brooklyn --password S3CR3T \
      CLOUD_USER_ACCESS_ID:CLOUD_SECRET@aws-ec2:us-west-1 > /tmp/last_container
    docker logs -f `cat /tmp/last_container`

then play with it at localhost:8081 or:

    export BR="docker -i exec `cat /tmp/last_container` br"
    $BR apps

(and the nicest thing is that you don't need anything installed or built locally, as long as you've got docker)

ahgittin commented:
build failure looks unrelated:

```
GitHub pull request #56 to apache/brooklyn-dist
[EnvInject] - Loading node environment variables.
ERROR: SEVERE ERROR occurs
org.jenkinsci.lib.envinject.EnvInjectException: hudson.remoting.ChannelClosedException: channel is already closed
	at org.jenkinsci.plugins.envinject.service.EnvironmentVariablesNodeLoader.gatherEnvironmentVariablesNode(EnvironmentVariablesNodeLoader.java:75)
```

neykov commented:
rat check failing with:
```
Unapproved licenses:

  docker/README.md
  docker/Makefile
```

mikezaccardo commented on docker/Dockerfile:
```diff
@@ -0,0 +1,59 @@
+
+USER brooklyn
+
+# handle customisation, including passing args through to the container script and installing boms
```
Funny coming from me, given that I pushed for HL-specificity for the first release, but I agree that the app-agnostic work should live here and the HL-specific child `Dockerfile` should live elsewhere.

ahgittin replied:
i thought that originally but the code for each project is basically the same, so it felt better to parametrise it here

neykov commented on docker/Dockerfile:
```diff
@@ -0,0 +1,59 @@
+FROM openjdk:8-jre-alpine
+MAINTAINER Alex Heneveld "alex@cloudsoft.io"
+
+RUN apk add --update bash openssh && rm -rf /var/cache/apk/*
```
Use `--no-cache` instead of `rm`-ing the cache.

neykov commented on docker/Dockerfile:
```diff
@@ -0,0 +1,60 @@
+MAINTAINER Alex Heneveld "alex@cloudsoft.io"
+
+RUN apk add --update --no-cache bash openssh ; \
+    mv /dev/random /dev/random-real && ln -s /dev/urandom /dev/random
```
Don't think this will work, have you checked?
Shouldn't be needed for docker anyway as it's using hosts entropy unlike VMs.

ahgittin replied:
agreed, and confirmed, the mods to `/dev` are lost in the next container (though it works if you do it in the same container).  also note you can't modify the `/proc` FS.  agree the answer is to expect the host to be configured correctly.  (often in clouds it isn't by default, still, including where i'm running.)

removed this

andreaturli commented on docker/Dockerfile:
```diff
@@ -0,0 +1,68 @@
+FROM openjdk:8-jre-alpine
+MAINTAINER Alex Heneveld "alex@cloudsoft.io"
+
+RUN apk add --update --no-cache bash openssh ; \
```
sorry I forgot to say `wget`

neykov commented:
Thanks for the updates, merging.

neykov commented on docker/Makefile:
```diff
@@ -0,0 +1,76 @@
+# under the License.
+#
+
+BROOKLYN_VERSION=0.10.0-SNAPSHOT
```
Just saw that - can you update and add `BROOKLYN_VERSION` tag.
Also do we need to keep `Dockerfile-debug-ubuntu`? It will get out of sync sooner or later so better remove it.

neykov replied:
bump @ahgittin

ahgittin commented:
Thanks @neykov -- changes addressed

neykov commented:
thanks @ahgittin, merging
